auth/store/mongo/device_tokens_repository.go

```diff
@@ -0,0 +1,53 @@
+package mongo
+
+import (
+	"context"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+
+	"github.com/tidepool-org/platform/devicetokens"
+	"github.com/tidepool-org/platform/errors"
+	structuredmongo "github.com/tidepool-org/platform/store/structured/mongo"
+)
+
+// deviceTokenRepo implements devicetokens.Repository, writing data to a
+// MongoDB collection.
+type deviceTokenRepo structuredmongo.Repository
+
+// Upsert will create or update the given Config.
+func (r *deviceTokenRepo) Upsert(ctx context.Context, doc *devicetokens.Document) error {
+	// The presence of UserID and TokenID should be enforced with a mongodb
+	// index, but better safe than sorry.
+	if doc.UserID == "" {
+		return errors.New("UserID may not be empty")
+	}
+	if doc.TokenKey == "" {
+		return errors.New("TokenID may not be empty")
+	}
+
+	opts := options.Update().SetUpsert(true)
+	f := bson.M{"tokenKey": doc.TokenKey, "userId": doc.UserID}
+	_, err := r.UpdateOne(ctx, f, bson.M{"$set": doc}, opts)
+	if err != nil {
+		return errors.Wrap(err, "upserting device token")
+	}
+	return nil
+}
+
+// EnsureIndexes to maintain index constraints.
+func (r *deviceTokenRepo) EnsureIndexes() error {
+	repo := structuredmongo.Repository(*r)
+	return (&repo).CreateAllIndexes(context.Background(), []mongo.IndexModel{
+		{
+			Keys: bson.D{
+				{Key: "userId", Value: 1},
+				{Key: "tokenKey", Value: 1},
+			},
+			Options: options.Index().
+				SetUnique(true).
+				SetName("UserIdTokenKeyTypeUnique"),
```

Comment:
It was a typo, yes. However, as this is live on qa2 now, I'd prefer to leave it. I don't think there's any chance of anyone mistaking it for something else, and it's preferable to having to muck with indexes, even if it is just a QA env.

Reply:
Hmm, well, I disagree since if I saw that index name used someplace I'd assume it had the `type` key in it. @toddkazakov what do you think here?

Reply:
Would you think it had `Token` or `IdToken` or `TokenKey` or maybe `UserIdToken`? Either way, you have to read the code, right?

Reply:
While I don't insist on specifying the keys in the index name (it's redundant), if this is the convention we are following, we shouldn't have fields in the name that are not there. I would have expected this index to include a field named `key`.

It's better to fix this now while the PR is in QA - we can just drop the index and it will be created automatically. If this was already in use in production I would have preferred to avoid changing the index name.

Reply:
Will do